### PR TITLE
My_Enumerables - Milestone -Feature branch

### PR DESCRIPTION
**Pull request by Rumbie-Mudzie7**
In this Milestone we:
-  Created our methods to replicate the original Enumerable methods.
-  Implemented DRY principles.

### CODE REVIEW

File: enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+  def my_each_with_index
+    return enum_for(:my_each) unless block_given?
+
+    array = is_a?(Array) ? self : to_a
+    i = 0
+    while i < array.length
+      yield(array[i], i)
+      i += 1
+    end
+  end
```
Comment (maelfosso) [inline enumerables.rb]:
```rb
2.6.5 :011 > array
 => [5, 7, 0, 7, 6, 0, 8, 0, 1, 6, 4, 5, 7, 6, 8, 4, 2, 6, 4, 5, 8, 5, 5, 0, 5, 1, 0, 6, 4, 2, 6, 3, 6, 8, 6, 0, 8, 3, 4, 5, 3, 0, 1, 2, 2, 5, 4, 3, 0, 8, 6, 7, 8, 4, 6, 2, 6, 2, 6, 7, 3, 4, 8, 7, 4, 7, 1, 3, 6, 7, 8, 1, 5, 5, 6, 2, 7, 5, 3, 1, 7, 8, 5, 6, 8, 0, 0, 6, 4, 8, 2, 7, 8, 0, 4, 6, 8, 6, 8, 4] 
2.6.5 :012 > block = proc { |num| num < 4 }
 => #<Proc:0x000000000254f518@(irb):12> 
2.6.5 :013 > array.my_each_with_index(&block)
 => nil 
2.6.5 :014 > array.each_with_index(&block)
 => [5, 7, 0, 7, 6, 0, 8, 0, 1, 6, 4, 5, 7, 6, 8, 4, 2, 6, 4, 5, 8, 5, 5, 0, 5, 1, 0, 6, 4, 2, 6, 3, 6, 8, 6, 0, 8, 3, 4, 5, 3, 0, 1, 2, 2, 5, 4, 3, 0, 8, 6, 7, 8, 4, 6, 2, 6, 2, 6, 7, 3, 4, 8, 7, 4, 7, 1, 3, 6, 7, 8, 1, 5, 5, 6, 2, 7, 5, 3, 1, 7, 8, 5, 6, 8, 0, 0, 6, 4, 8, 2, 7, 8, 0, 4, 6, 8, 6, 8, 4] 
2.6.5 :015 > 
```

File: enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    selected
+  end
+
+  # My_all?
+  def my_all?(args = nil)
+    my_select { |el| return false unless el.match?(args) } if args.class == Regexp
+    my_select { |el| return false unless el.is_a?(args) } if args.class == Class
+    !block_given? ? my_select { |item| return false unless item } : my_select { |item| return false unless yield(item) }
+    true
+  end
```
Comment (maelfosso) [inline enumerables.rb]:
```rb
2.6.5 :017 > array
 => [5, 7, 0, 7, 6, 0, 8, 0, 1, 6, 4, 5, 7, 6, 8, 4, 2, 6, 4, 5, 8, 5, 5, 0, 5, 1, 0, 6, 4, 2, 6, 3, 6, 8, 6, 0, 8, 3, 4, 5, 3, 0, 1, 2, 2, 5, 4, 3, 0, 8, 6, 7, 8, 4, 6, 2, 6, 2, 6, 7, 3, 4, 8, 7, 4, 7, 1, 3, 6, 7, 8, 1, 5, 5, 6, 2, 7, 5, 3, 1, 7, 8, 5, 6, 8, 0, 0, 6, 4, 8, 2, 7, 8, 0, 4, 6, 8, 6, 8, 4] 
2.6.5 :018 > array.my_all?(3)
 => true 
2.6.5 :019 > array.all?(3)
 => false 
2.6.5 :020 > 
```

File: enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+  # My_any
+  def my_any?(args = nil)
+    if args.class == Regexp
+      ar = []
+      my_each { |el| ar << el if el.match?(args) }
+      return ar.empty? ? false : true
+    end
```
Comment (maelfosso) [inline enumerables.rb]:
```rb
2.6.5 :025 > false_array = [nil, false, nil, false]
 => [nil, false, nil, false] 
2.6.5 :026 > false_array.any?
 => false 
2.6.5 :027 > false_array.my_any?
Traceback (most recent call last):
        9: from /usr/share/rvm/rubies/ruby-2.6.5/bin/irb:23:in `<main>'
        8: from /usr/share/rvm/rubies/ruby-2.6.5/bin/irb:23:in `load'
        7: from /usr/share/rvm/rubies/ruby-2.6.5/lib/ruby/gems/2.6.0/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        6: from (irb):27
        5: from enumerables.rb:64:in `my_any?'
        4: from enumerables.rb:40:in `my_select'
        3: from enumerables.rb:9:in `my_each'
        2: from enumerables.rb:40:in `block in my_select'
        1: from enumerables.rb:64:in `block in my_any?'
LocalJumpError (no block given (yield))
2.6.5 :028 > 
```

Comment (maelfosso) [reply enumerables.rb]:
```rb
2.6.5 :034 > words
 => ["dog", "door", "rod", "blade"] 
2.6.5 :035 > words.any?(Integer)
 => false 
2.6.5 :036 > words.my_any?(Integer)
 => true 
2.6.5 :037 > 
```

Comment (maelfosso) [reply enumerables.rb]:
```rb
2.6.5 :040 > words
 => ["dog", "door", "rod", "blade"] 
2.6.5 :041 > words.my_any?('cat')
 => true 
2.6.5 :042 > words.any?('cat')
 => false 
2.6.5 :043 > 
```

File: enumerables.rb [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+      return ar.empty? ? true : false
+    end
+    if args.class == Class
+      ar = []
+      my_each { |el| ar << el if el.is_a?(args) }
+      return ar.empty? ? true : false
+    end
+    !block_given? ? my_select { |item| return false if item } : my_select { |item| return false if yield(item) }
+    true
+  end
```
Comment (maelfosso) [inline enumerables.rb]:
```rb
2.6.5 :044 > words
 => ["dog", "door", "rod", "blade"] 
2.6.5 :045 > words.my_none?(5)
 => false 
2.6.5 :046 > words.none?(5)
 => true 
2.6.5 :047 >
```

File: enumerables.rb [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+  # MY_MAP
+  def my_map(proc = nil)
+    ar = []
+    my_each { |el| ar << proc.call(el) } if proc
+    return enum_for(:my_map) unless block_given?
+
+    my_each { |el| ar << yield(el) }
+    ar
+  end
```
Comment (maelfosso) [inline enumerables.rb]:
```rb
2.6.5 :049 > my_proc = proc {|num| num > 10 }
 => #<Proc:0x000000000239d8c8@(irb):49> 
2.6.5 :050 > array.my_map(my_proc) {|num| num < 10 }
 => [false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true] 
2.6.5 :051 > array.map(&my_proc)
 => [false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false] 
2.6.5 :052 > array.my_map(my_proc) {|num| num < 10 } = array.map(&my_proc)
Traceback (most recent call last):
        3: from /usr/share/rvm/rubies/ruby-2.6.5/bin/irb:23:in `<main>'
        2: from /usr/share/rvm/rubies/ruby-2.6.5/bin/irb:23:in `load'
        1: from /usr/share/rvm/rubies/ruby-2.6.5/lib/ruby/gems/2.6.0/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
SyntaxError ((irb):52: syntax error, unexpected '=', expecting end-of-input)
...ap(my_proc) {|num| num < 10 } = array.map(&my_proc)
...                              ^
2.6.5 :053 > array.my_map(my_proc) {|num| num < 10 } == array.map(&my_proc)
 => false 
2.6.5 :054 > 
```

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello @Rumbie-Mudzie7 @vmwhoami 

Thank you for that _almost_ well done job.
To make it completely well done, there are some bugs to fix.
Kindly read the comments below please.

**PS**: Please, it's `my_inject` not **my_reduce** as you did. Please, update it.

Once done, submit a new code review request.
Thank you

Mael FOSSO

**Comment by lekegitrepo** [comment]
#### Hi @Rumbie-Mudzie7 and @vmwhoami :clap :clap:
#### Well done. You have done a good job :+1: However `my_map` method is not working, fix it and also change `my_reduce` call in `multiply_els` to `my_inject`.

**_Fix this and submit for review._**

##### _Reviewed by Temitayo Adeleke_.

**Review by lekegitrepo — CHANGES_REQUESTED** [review]
##### Status: Request changes

**Comment by lekegitrepo** [comment]
#### Good job :clap: :clap:
##### Well done. This project is approved :star: :star:

#### _Reviewed by Temitayo Adeleke_.

**Review by lekegitrepo — APPROVED** [review]
#### Status: Approve

**Pull request merged**